### DISCUSSION TITLE
Setting differential learning rates

### DISCUSSION THREAD

**Discussion by jcpayne** [facebookresearch/detectron2#2604] [Q&A]
I'm just wondering if the ability to set different learning rates for parameter groups within the model is implemented in Detectron2; for example could you train three parameter groups where the bottom layers of the model are training slower than the top, with `lrs = np.array([lr/9, lr/3, lr]`?  

If I'm reading the code right, it looks like `build_optimizer()` and `build_lr_scheduler()` both are both set up to use a single learning rate for all of the layers in a model, but then WarmupMultiStepLR and WarmupCosineLR both have a variable that holds a list of learning rates (`self.base_lrs`).  

I may have missed it, but I don't see where you would set the parameter groups or the corresponding learning rates.

Thanks.

**Comment by ppwwyyxx** [comment]
You can use any custom optimizer (https://detectron2.readthedocs.io/en/latest/tutorials/training.html), so you can set param groups to whatever needed.

https://github.com/facebookresearch/detectron2/blob/c152862a18182544b9f41dac0519e87c5ff8ac6c/projects/Panoptic-DeepLab/train_net.py#L111-L112 is one example

**Comment by jcpayne** [comment]
Sorry to be dumb but where would you define the parameter groups and where would you set the corresponding learning rates?

**Reply by ppwwyyxx** [reply]
That's about pytorch optimizer and not detectron2, see https://pytorch.org/docs/stable/optim.html for some docs

**Reply by jcpayne** [reply]
OK thanks for the pointer.  I'll check it out.
